snapshot/volume-snapshotting-proposal.md

```diff
@@ -0,0 +1,421 @@
+	PersistentVolumeRef *core_v1.ObjectReference
+}
+
+// Represents the actual location and type of the snapshot. Only one of its members may be specified.
+type VolumeSnapshotDataSource struct {
+    // Per-plugin specific spec
+}
+```
+
+* Event loop in controller
+
+The volume snapshot controller maintains two data structures (ActualStateOfWorld and DesiredStateOfWorld) and
+periodically reconciles the two. The data structures are being update by the API sever event handlers.
+
+* If a new `VolumeSnapshot` is added, the *add* handler adds it to the DesiredStateOfWorld (DSW)
+
+* If a `VolumeSnapshot` is deleted, the *delete* handler removes it from the DSW.
+
+* Reconciliation loop in the controller
+
+    * For every `VolumeSnapshot` in the ActualStateOfWorld (ASW) find the corresponding `VolumeSnapshot` in the DSW.
+    If such a snapshot does not exist, start a snapshot deletion operation:
+
+        * Determine the correct volume snapshot plugin to use from the `VolumeSnapshotData` referenced by the
+        `VolumeSnapshot`
+
+        * Create a delete operation: only one such operation is allowed to exist for the given `VolumeSnapshot` and
+        `VolumeSnapshotData` pair
+
+        * The operation is an asynchronous function using the volume plugin to delete the actual snapshot in the back-end.
+
+        * When the plugin finishes deleting the snapshot, delete the `VolumeSnapshotData` referencing it and remove the
+        `VolumeSnapshot` reference from the ASW
+
+    * For every `VolumeSnapshot` in the DSW find the corresponding `VolumeSnapshot` in the ASW. If such a snapshot
+    does not exist, start a snapshot creation operation:
+
+        * Determine the correct volume snapshot plugin to use from the `VolumeSnapshotData` referenced by the
+        `VolumeSnapshot`
+
+        * Create a volume snapshot creation operation: only one such operation is allowed to exist for the given
+        `VolumeSnapshot` and `VolumeSnapshotData` pair.
+
+        * The operation is an asynchronous function using the volume plugin to create the actual snapshot in the back-end.
+
+        * When the plugin finishes creating the snapshot a new `VolumeSnapshotData` is created holding a reference to
+        the actual volume snapshot.
+
+    * For every snapshot present in the ASW and DSW find its `VolumeSnapshotData` and verify the bi-directional
+    binding is correct: if not, update the `VolumeSnapshotData` reference.
+
+
+### Create Snapshot Logic
+
+To create a snapshot:
+
+* Acquire operation lock for volume so that no other attach or detach operations can be started for volume.
+
+    * Abort if there is already a pending operation for the specified volume (main loop will retry, if needed).
+
+* Spawn a new thread:
+
+    * Execute the volume-specific logic to create a snapshot of the persistent volume referenced by the PVC.
+
+    * For any errors, log the error, and terminate the thread (the main controller will retry as needed).
```

Comment:
+ send the errors as events on corresponding `VolumeSnapshot` object so user can see what's going on in `kubectl describe volumesnapshot`